msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs

```diff
@@ -0,0 +1,153 @@
+            {
+                ClientId = TestConstants.ClientId
+            };
+
+            AuthenticationRequestParameters parameters = new AuthenticationRequestParameters()
+            {
+                Authority = authority,
+                // todo: what is this? SliceParameters = "key1=value1%20with%20encoded%20space&key2=value2",
+                ClientId = TestConstants.ClientId,
+                Scope = TestConstants.Scope,
+                TokenCache = _cache,
+                RequestContext = new RequestContext(new MsalLogger(Guid.NewGuid(), null))
+            };
+
+            RequestTestsCommon.MockInstanceDiscoveryAndOpenIdRequest();
+
+            var expectedScopes = new HashSet<string>();
+            expectedScopes.UnionWith(TestConstants.Scope);
+            expectedScopes.Add(OAuth2Value.ScopeOfflineAccess);
+            expectedScopes.Add(OAuth2Value.ScopeProfile);
+            expectedScopes.Add(OAuth2Value.ScopeOpenId);
+
+            // Mock Handler for device code request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = CreateDeviceCodeResponseSuccessMessage()
+            });
+
+            // Mock Handler for devicecode->token exchange request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = MockHelpers.CreateSuccessTokenResponseMessage()
+            });
+
+            DeviceCodeResult actualDeviceCodeResult = null;
+            DeviceCodeRequest request = new DeviceCodeRequest(parameters, result => 
+            {
+                actualDeviceCodeResult = result;
+                return Task.FromResult(0);
+            });
+            var task = request.RunAsync(CancellationToken.None);
+            task.Wait();
+            var authenticationResult = task.Result;
+            Assert.IsNotNull(authenticationResult);
+            Assert.IsNotNull(actualDeviceCodeResult);
+
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+            Assert.AreEqual("BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA", actualDeviceCodeResult.DeviceCode);
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);

```

Comment:
>Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId); [](start = 12, length = 62)

you have this test above. did you intend to make this test 2 times?

Reply:
Fixed, wired in scopes.  Also added test for cancel.